pkgs/by-name/ko/koboldcpp/package.nix

```
@@ -0,0 +1,98 @@
+{ lib
+, fetchFromGitHub
+, stdenv
+, autoPatchelfHook
+, makeWrapper
+, python311Packages
+, tk
+
+, config
+, cudaPackages ? { }
+
+, openblasSupport ? true
+, openblas
+
+, cublasSupport ? config.cudaSupport
+
+, clblastSupport ? true
+, clblast
+, ocl-icd
+
+, vulkanSupport ? true
+, vulkan-loader
+
+, metalSupport ? stdenv.isDarwin && stdenv.isAarch64 && !openblasSupport
+}:
+
+let
+  inherit (lib) optionals;
+
+  pythonInputs = with python311Packages; [
+    tkinter
+    customtkinter
+    packaging
+    tk
+  ];
+
+  cudaBuildInputs = with cudaPackages; [
+      libcublas
+  ];
+in
+stdenv.mkDerivation (finalAttrs: {
+  pname = "koboldcpp";
+  version = "1.65";
+
+  src = fetchFromGitHub {
+    owner = "LostRuins";
+    repo = "koboldcpp";
+    rev = "refs/tags/v${finalAttrs.version}";
+    hash = "sha256-GYvU+0bhskCgSXAoeSy/2wWIF5FrilLVdXitFvuUAdQ=";
+  };
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+    makeWrapper
+    python311Packages.wrapPython
+  ];
+
+  buildInputs = pythonInputs ++ optionals openblasSupport [ openblas ]
+  ++ optionals cublasSupport [ cudaPackages.libcublas ]
+  ++ optionals clblastSupport [ clblast ocl-icd ]
+  ++ optionals vulkanSupport [ vulkan-loader ];
+
+  pythonPath = pythonInputs;
+
+  makeFlags = optionals openblasSupport [ "LLAMA_OPENBLAS=1" ]
+  ++ optionals cublasSupport [ "LLAMA_CUBLAS=1" ]
+  ++ optionals clblastSupport [ "LLAMA_CLBLAST=1" ]
+  ++ optionals vulkanSupport [ "LLAMA_VULKAN=1" ]
+  ++ optionals metalSupport [ "LLAMA_METAL=1" ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/run-env
+
+    install -Dm755 koboldcpp.py $out/run-env/koboldcpp-unwrapped
+    cp *.so $out/run-env
```

```suggestion
    mkdir -p "$out/run-env"

    install -Dm755 koboldcpp.py "$out/run-env/koboldcpp-unwrapped"
    cp *.so "$out/run-env"
```

You should move all `*.embd` and `quantize_*` files as well

---

The reason I did `run-env` is that koboldcpp needs the koboldcpp_default.so file to be in the same directory.

I wasn't sure if it was proper to leave the shared object file in `bin` so I put them both into a separate directory and put the wrapper in `bin`. But I don't think this is supposed to be the proper solution here.

---

You should use `$out/bin` instead of `$out/run-env` as there is no mention or usage of `run-env` in the manual or anywhere on nixpkgs

All `embd` and `so` files should be in same directory as the `koboldcpp.py`